flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/join/stream/state/MultiJoinStateViews.java

```diff
@@ -0,0 +1,440 @@
+
+            // Composite key type: RowData with 2 fields (joinKey, uniqueKey)
+            // The composite key is a RowData with joinKey at index 0 and uniqueKey at index 1.
+            // TODO Gustavo: there is probably a cleaner way of instantiating the type
+            final RowType keyRowType =
+                    RowType.of(joinKeyType.toRowType(), uniqueKeyType.toRowType());
+            InternalTypeInfo<RowData> compositeKeyType = InternalTypeInfo.of(keyRowType);
+
+            MapStateDescriptor<RowData, RowData> recordStateDesc =
+                    createStateDescriptor(stateName, compositeKeyType, recordType, ttlConfig);
+
+            this.recordState = ctx.getMapState(recordStateDesc);
+        }
+
+        private RowData createCompositeKey(RowData joinKey, RowData uniqueKey) {
+            GenericRowData compositeKey = new GenericRowData(2);
+            compositeKey.setField(0, joinKey);
+            compositeKey.setField(1, uniqueKey);
+            return compositeKey;
+        }
+
+        @Override
+        public void addRecord(RowData joinKey, RowData record) throws Exception {
+            RowData uniqueKey = uniqueKeySelector.getKey(record);
+            RowData compositeKey = createCompositeKey(joinKey, uniqueKey);
+            recordState.put(compositeKey, record);
+        }
+
+        @Override
+        public void retractRecord(RowData joinKey, RowData record) throws Exception {
+            RowData uniqueKey = uniqueKeySelector.getKey(record);
+            RowData compositeKey = createCompositeKey(joinKey, uniqueKey);
+            recordState.remove(compositeKey);
+        }
+
+        @Override
+        public Iterable<RowData> getRecords(RowData joinKey) throws Exception {
+            Iterator<Map.Entry<RowData, RowData>> stateIterator = recordState.iterator();
+            if (stateIterator == null) {
+                return Collections.emptyList();
+            }
+
+            // Consume the record
+            // This iterator is stateful and intended for single use per call to getRecords.
+            //noinspection NullableProblems TODO Gustavo can we not supress?
+            return new IterableIterator<>() {
+                private RowData nextRecord = null;
+
+                @Override
+                public boolean hasNext() {
+                    if (nextRecord != null) {
+                        return true;
+                    }
+                    while (stateIterator.hasNext()) {
+                        Map.Entry<RowData, RowData> entry = stateIterator.next();
+                        RowData compositeKey = entry.getKey();
+                        RowData currentJoinKey = compositeKey.getRow(0, joinKeyType.getArity());
+                        if (joinKey.equals(currentJoinKey)) {
+                            nextRecord = entry.getValue();
+                            return true;
+                        }
+                    }
+                    return false;
```

Comment:
Can we terminate the iteration as soon as the record doesn't match join condition? I don't think so.
IIUC, `recordState` might look like this:
```
// JK, UK, value
 <0,   a,    ...>
 <1,   b,     ...>
 <1,   c,     ...>
 <2,   d,     ...>

(this is already with some common key selected)

So for JoinKey 1, we need to collect two records in the middle.
(also, there're no guarantees that the records are sorted in general)

Or am I missing something?

Reply:
I'm pretty sure that this optimization would be required, but I still might be wrong :)
It would be great to add a benchmark to https://github.com/apache/flink-benchmarks and compare the performance of the new approach with the current one.
This might also be helpful to judge about https://github.com/apache/flink/pull/26313#discussion_r2105804426

Reply:
> I'm pretty sure that this optimization would be required, but I still might be wrong :)

How could we implement this RocksDB-optimized version - is this an operator decision? If you have any pointers to where we do something similar, would be happy to look into those.

> It would be great to add a benchmark to https://github.com/apache/flink-benchmarks

Yes, there are multiple smaller optimizations that I have in my todo list to try out but wanted to do them only after I'm able to measure things. Basically following one of the "general remarks" from the benchmarks repo "If you can not measure the performance difference, then just don't bother (avoid premature optimisations)"

Reply:
Ok, I've looked into this. In short, we'll need a SortedMapState to implement this optimization. With that, we'd be able to seek the first matching keys and also exit early when we have no more matching join keys. This is definitely a considerable optimization.

We'd need something like this https://cwiki.apache.org/confluence/display/FLINK/FLIP-220%3A+Binary+Sorted+State and could revisit as soon as we have it.